### PR TITLE
feat(sso): Multitenant SAML

### LINKED ISSUE #5647: Full SAML login support
**EDIT: Option A below is implemented. When it makes sense to prioritize we should consider multi-tenancy SAML (Option B below).**

------- 

## Option B. Organization-based. Self-hosted + Cloud.
- You configure SAML login on an organization-basis. Up to one SAML provider per organization.
- The login process could then look something like this:
    1. We'd have to refactor the login page so you enter your email first and based on your email we show you the appropriate login options.
    2. The above could raise a security concern about potentially exposing an endpoint you could use to find out if a certain user has an account on our system.
    3.  After the user enters their email address (from [1]), if they only have a single SSO login option, we take them to the respective login page.
    4. If the user has multiple login options (e.g. password + SSO), we let them choose which option they want (like we currently do in our login page).
- This option also brings up a few more edge cases.
    - As an org admin I probably want to restrict my users to only have my SAML login option (otherwise the purpose of SAML would be defeated), but a user can belong to multiple organizations, so we need to give the option for either a) users in your org can only login with your SSO and/or b) users in your org can only belong to your org.
    - On self-hosted we would want to still take the users to their login page automatically if they have a single org and a single authentication mechanism.
    - Because a user can technically belong to many orgs each with its own auth mechanism, we could technically end up with users with multiple SAML authentication paths, we would have to support each path.

-------

## Additional context from original issue

<details>
 <summary>Expand for additional context</summary>
 Addresses #5547. Some focus customers and a few other users have requested support for SAML. This is a compliance feature that a lot of large-scale organizations require for full widespread rollout. When working at a large org, having a centralized place for provisioning and managing users can make the onboarding/offboarding process smoother and help reduce security risks. There's a great [Slack thread](https://posthogusers.slack.com/archives/CT7HXDEG3/p1628859965082200) by a community member about the importance of SAML and [this study](https://www.onelogin.com/press-center/press-releases/97-percent-of-saas-vendors-backing-saml-based-single-sign-on) by OneLogin in which they find that up to 97% of SaaS support have support (or plans) for SAML.

Aside from the implementation complexities of SAML (it's a more robust standard, so it's not as straightforward as our current social auth providers), there's some conceptual difficulties of implementing SAML.

## Option A. Instance-based. Only Self-hosted.
- We enable instance-based configuration for SAML. You can have only one SAML provider.
- We give users the option to only allow SAML login to their instance (and restrict password creation & access), in this case the default login screen fully disappears and we always redirect to their login page.
- As a fallback when something breaks down, admins could simply disable SAML for their instance (we could even consider giving automatic Django admin access to allow them to reset passwords).

## Final notes

**Which option to choose?** Prioritizing this solely for solving for our focus customers I would go with Option A, not only based on our current customers but based on our persona profile (probably all of these will self-host). It's important to consider that based on the referenced study above and how other SaaS operate, it seems we will eventually require this on Cloud, but I would argue we might be better off pushing this down the line until it's actually needed (opportunity cost of other stuff with more immediate impact), but let's not forget this will require a loan from our tech debt bank.

</details>

### PR DESCRIPTION
**Pull request by paolodamico**
## Problem

It's finally here 👮 ! Closes #5647. @camerondeleone can provide more context from the customer side.

## Changes
- Implements multitenancy SAML support, paritcularly aimed at enabling SAML on PostHog Cloud.
- Allows enforcing SAML and just-in-time provisioning.
- SAML requires an Enterprise license and SSO enforcement too (treated as separate features in case we want to offer one or the other on future plans).
- If a domain is claimed and verified but doesn't have just-in-time provisioning, we won't create a new organization for them when logging in through an SSO provider (which has caused confusion in the past).
- As SAML is now the only non-instance-based SSO provider, you can no longer start the SAML flow from the login page **without entering your email address first.**
- ⚠️ Introduces major deprecation for self-hosted users. SAML will no longer be configured instance-wide via env vars. Admin users will have to add their SAML config to a specific domain. 

## How did you test this code?
- Pytest, pytest, pytest.

### CODE REVIEW

**Comment by paolodamico** [comment]
@mariusandra apologies if this PR is a bit larger, I tried to split this massive project into as many chunks as possible to facilitate review. You can use OneLogin / Okta credentials in 1Password to test the real deal (I'm doing a test right now).

**Comment by paolodamico** [comment]
FYI just did a complete test using OneLogin and everything works as expected 🙌! Please be mindful that we need some minor changes in the frontend (separate PR coming) for this to work 100%

File: ee/settings.py [hunk 1]
```diff
@@ -17,8 +17,13 @@
 HOOK_FINDER = "ee.models.hook.find_and_fire_hook"
 HOOK_DELIVERER = "ee.models.hook.deliver_hook_wrapper"
 
-# SAML
-SAML_CONFIGURED = False
+# SSO
+AUTHENTICATION_BACKENDS = AUTHENTICATION_BACKENDS + [
```
Comment (mariusandra) [inline ee/settings.py:21]: Python is funny

**Review by mariusandra — APPROVED** [review]
Looks very good!

**Comment by mariusandra** [comment]
I did not run a full test, but the code looks good. Are there any blockers for merging this (and [the docs](https://github.com/PostHog/posthog.com/pull/3245)) in before the release @timgl ? Should it make it into [the release branch](https://github.com/PostHog/posthog/tree/release-1.34.0)?

**Comment by paolodamico** [comment]
Hmm not sure if we want to include in this release. If we do, we may want to 1) do a quick pass of the docs update to make sure any references to 1.35.0 are updated to 1.34.0; 2) update the Array with the appropriate deprecation notices; 3) do a full e2e test on playground.

**Comment by mariusandra** [comment]
Let's get this in and see what happens 🎉

**Comment by camerondeleone** [comment]
I'm assuming there is still some UI updating in PostHog/posthog-cloud that we need to reveal SAML as an auth option on app.posthog.com, since I still see the standard login and social auth only.

**Comment by mariusandra** [comment]
We need to set it up for it to show up:
image

Each organisation can set it up for their own domain with their own settings.

**Comment by paolodamico** [comment]
Exactly as @mariusandra said. Just to add to that, the button will not appear immediately as it does for Google, GitHub, ..., but rather when you type an email address it will appear if SAML is configured for that email address.

**Pull request merged**